zkvm/src/signer/mod.rs

```diff
@@ -0,0 +1,229 @@
+#![allow(non_snake_case)]
+
+use crate::transcript::TranscriptProtocol;
+use curve25519_dalek::ristretto::RistrettoPoint;
+use curve25519_dalek::scalar::Scalar;
+use merlin::Transcript;
+use rand;
+
+// Modules for signing protocol
+
+pub mod prover;
+
+#[derive(Clone)]
+pub struct PrivKey(Scalar);
+
+#[derive(Clone)]
+pub struct PubKey(RistrettoPoint);
+
+pub struct MultiKey(Vec<PubKey>); // TODO: also include Option<Scalar> for signing key?
+
+#[derive(Clone)]
+pub struct PubKeyHash(Scalar);
+
+#[derive(Debug, Clone)]
+pub struct Signature {
+    s: Scalar,
+    R: RistrettoPoint,
+}
+
+#[derive(Clone)]
+pub struct Shared {
+    G: RistrettoPoint,
+    transcript: Transcript,
+    // X_agg = sum_i ( a_i * X_i )
+    X_agg: PubKey,
+    // L = H(X_1 || X_2 || ... || X_n)
+    L: PubKeyHash,
+    // message being signed
+    m: Vec<u8>,
+}
+
+impl MultiKey {
+    pub fn aggregate(&self, transcript: &mut Transcript) -> (PubKey, PubKeyHash) {
```

Comment:
Aggregation is the "entry-point" w.r.t. the transcript (unlike signing which is composable with whatever prior ZKP operations) — so instead of passing it in, we need to just hardcode the label like `Transcript::new("ZkVM.aggregated-key")`.

Reply:
should we return the trancript from this function then?

Reply:
nope. It's an "internal hash" instance only.

Reply:
Hmm. the hardcoded label used in the `aggregate` function HAS to match the label in the signer states (since otherwise, they would get different `a_i`, `c_i` challenges). So, I'm guessing that means we should also hardcode a transcript state into `PartyAwaitingPrecommitments::new()` as well

Reply:
Resolved - we are hardcoding the transcript for pubkey aggregation, and passing in the transcript for nonce commitment generation, factor_for_key generation etc (basically, for prover and verifier use)